solidity/test/token_stake/TestPunishment.js

```diff
@@ -0,0 +1,209 @@
+
+    before(async () => {
+        token = await KeepToken.new({ from: owner })
+        registry = await Registry.new({ from: owner })
+        stakingContract = await TokenStaking.new(
+            token.address,
+            registry.address,
+            initializationPeriod,
+            undelegationPeriod,
+            { from: owner }
+        )
+
+        await registry.setRegistryKeeper(registryKeeper, { from: owner })
+
+        minimumStake = await stakingContract.minimumStake()
+        largeStake = minimumStake.muln(2)
+
+        await registry.approveOperatorContract(
+            operatorContract,
+            { from: registryKeeper }
+        )
+
+        await stakeDelegate(
+            stakingContract, token, owner, operator,
+            owner, authorizer, largeStake
+        )
+
+        await stakingContract.authorizeOperatorContract(
+            operator,
+            operatorContract,
+            { from: authorizer }
+        )
+    });
+
+    beforeEach(async () => {
+        await createSnapshot()
+    })
+
+    afterEach(async () => {
+        await restoreSnapshot()
+    })
+
+    describe("slash", () => {
+        it("should slash token amount from stake", async () => {
+            time.increase((await stakingContract.initializationPeriod()).addn(1))
+
+            let amountToSlash = web3.utils.toBN(42000000);
+
+            let balanceBeforeSlashing = await stakingContract.balanceOf(operator)
+            await stakingContract.slash(amountToSlash, [operator], { from: operatorContract })
+            let balanceAfterSlashing = await stakingContract.balanceOf(operator)
+
+            expect(balanceAfterSlashing).to.eq.BN(balanceBeforeSlashing.sub(amountToSlash))
+        })
+
+        it("should slash no more than available on stake", async () => {
+            time.increase((await stakingContract.initializationPeriod()).addn(1))
+
+            let amountToSlash = largeStake.add(web3.utils.toBN(100))
+            await stakingContract.slash(amountToSlash, [operator], { from: operatorContract })
+            let balanceAfterSlashing = await stakingContract.balanceOf(operator)
+
+            expect(balanceAfterSlashing).to.eq.BN(0)
+        })
+
+        it("should not fail if operator is slashed to zero", async () => {
+            time.increase((await stakingContract.initializationPeriod()).addn(1))
+
+            let amountToSlash = largeStake
+
+            // the first slash will slash to 0, the second one has nothing
+            // to slash; it should not fail
+            await stakingContract.slash(amountToSlash, [operator], { from: operatorContract })
+            await stakingContract.slash(amountToSlash, [operator], { from: operatorContract })
+
+            let balanceAfterSlashing = await stakingContract.balanceOf(operator)
+            expect(balanceAfterSlashing).to.eq.BN(0)
+        })
+
+        it("should fail when operator stake is not active yet", async () => {
```

Comment:
We should also test for a case where the stake is released. Slashing should fail, because it should not be trying to punish operators past the undelegation period without having placed a lock.

Reply:
Added in https://github.com/keep-network/keep-core/pull/1625/commits/6311811ae4d73aaeed76f261c0123d3094a7ca7f, I hade additionally covered the case it's not possible to slash/seize unlocked stake after undelegation in https://github.com/keep-network/keep-core/pull/1625/commits/439a5792a8915b13b94f64bcd0fefb4982ae1e0d.